client/network/src/request_responses.rs

```diff
@@ -445,30 +475,89 @@ impl NetworkBehaviour for RequestResponsesBehaviour {
 
 						// Received a response from a remote to one of our requests.
 						RequestResponseEvent::Message {
+							peer,
 							message:
 								RequestResponseMessage::Response {
 									request_id,
 									response,
 								},
 							..
 						} => {
+							let (started, delivered) = match self.pending_requests.remove(&request_id) {
+								Some((started, pending_response)) => {
+									let delivered = pending_response.send(
+										response.map_err(|()| RequestFailure::Refused),
+									).map_err(|_| RequestFailure::Obsolete);
+									(started, delivered)
+								}
+								None => {
+									log::warn!(
+										target: "sub-libp2p",
+										"Received `RequestResponseEvent::Message` with unexpected request id {:?}",
+										request_id,
+									);
+									debug_assert!(false);
+									continue;
+								}
+							};
+
 							let out = Event::RequestFinished {
-								request_id,
-								result: response.map_err(|()| RequestFailure::Refused),
+								peer,
+								protocol: protocol.clone(),
+								duration: started.elapsed(),
+								result: delivered,
 							};
+
 							return Poll::Ready(NetworkBehaviourAction::GenerateEvent(out));
 						}
 
 						// One of our requests has failed.
 						RequestResponseEvent::OutboundFailure {
+							peer,
 							request_id,
 							error,
 							..
 						} => {
+							// TODO: Remove hack by deriving `Clone` for `OutboundFailure`.
```

Comment:
I guess you intend to open a libp2p PR for it? There is no reason it couldn't be clone right now, though having to ensure errors are `Clone` can often be a nuisance that already starts with `io::Error` not being clonable. It is usually preferable if errors are propagated / moved only along a single code path. I wouldn't go so far to say that it is a code smell but it does feel a bit awkward to report finished requests over two separate code paths. The use of `oneshot` channels with their short-cut capacity to bypass all the layers that the `Sender` had to be passed through when it is ultimately given a value is very convenient and even efficient but it also makes all these layers unaware of what is sent through the `Sender`. This then seems to trigger the need to report these results _also_ separately. I don't have a better idea at the moment, just an impression that not having to clone the error in the first place would be preferable, i.e. not to report the same error over two different code paths.

Reply:
> I guess you intend to open a libp2p PR for it?

Correct.

> The use of oneshot channels with their short-cut capacity to bypass all the layers that the Sender had to be passed through when it is ultimately given a value is very convenient and even efficient but it also makes all these layers unaware of what is sent through the Sender. 

True. I think the usage of `oneshot` to bypass layers is an intrusive architectural pattern worth discussing. In my eyes, the reduced control is worth the simplification in this case. What do you think @romanb?

> I don't have a better idea at the moment, just an impression that not having to clone the error in the first place would be preferable, i.e. not to report the same error over two different code paths.

Agreed. A single code path would reduce the mental overhead. The two code paths are inherent to the fact that (a) layers are skipped via `oneshot` and (b) statistical reporting should be generic over all specific request response protocols.

One thing alleviating this problem: The error clone created here is used for statistic purpose only. Thus at least it is not ambiguous who needs to handle the error.

One approach where one would not need to clone the `OutboundFailure`, but still support the additional code path for statistics would be to derive a _statistics-specific_ error based on the `OutboundFailure` error. In my eyes this would not be worth the additional complexity (one more `struct`) but instead only hide the code-smell itself.

Reply:
> In my eyes, the reduced control is worth the simplification in this case. What do you think @romanb?

Pulling the `pending_responses` back one layer, and thus the use of these channels, does seem to be an improvement, yes. I guess my discomfort is more in the approach of reporting events between layers _only_ to collect metrics somewhere else. This has likely been discussed already and the decision made to prefer emitting events just to collect statistics elsewhere, and I know the trade-offs. I guess in my eyes metrics are a [cross-cutting concern](https://en.wikipedia.org/wiki/Cross-cutting_concern) like logging that is necessarily distributed throughout the code-base, except for the possible use of fancy techniques like aspect-oriented programming. Since I wouldn't send around events between different parts of the code just so I can have all the logging in a single place, I probably wouldn't do it for metrics collection either, but again YMMV.